vnext/ReactUWP/Views/Image/ReactImageBrush.cpp

```diff
@@ -0,0 +1,204 @@
+        surfaceBrush.Stretch(ResizeModeToStretch());
+
+        auto compositionBrush{ surfaceBrush.as<winrt::CompositionBrush>() };
+        if (ResizeMode() == ResizeMode::Repeat)
+        {
+          compositionBrush = GetOrCreateEffectBrush(surfaceBrush);
+        }
+
+        // Only switch CompositionBrush if we are switching to/from ResizeMode::Repeat
+        if (ShouldSwitchCompositionBrush())
+        {
+          CompositionBrush(compositionBrush);
+        }
+      }
+    }
+
+    winrt::CompositionEffectBrush ReactImageBrush::GetOrCreateEffectBrush(winrt::CompositionSurfaceBrush const& surfaceBrush)
+    {
+      if (!m_effectBrush)
+      {
+        auto borderEffect{ winrt::make<winrt::implementation::BorderEffect>() };
+        borderEffect.ExtendX(winrt::CanvasEdgeBehavior::Wrap);
+        borderEffect.ExtendY(winrt::CanvasEdgeBehavior::Wrap);
+
+        winrt::CompositionEffectSourceParameter borderEffectSourceParameter{ L"source" };
+        borderEffect.Source(borderEffectSourceParameter);
+
+        auto effectFactory{ winrt::Window::Current().Compositor().CreateEffectFactory(borderEffect) };
+        m_effectBrush = effectFactory.CreateBrush();
+
+        surfaceBrush.HorizontalAlignmentRatio(0.0f);
+        surfaceBrush.VerticalAlignmentRatio(0.0f);
+
+        m_effectBrush.SetSourceParameter(L"source", surfaceBrush);
+      }
+
+      return m_effectBrush;
+    }
+
+    winrt::CompositionSurfaceBrush ReactImageBrush::GetOrCreateSurfaceBrush()
+    {
+      // If it doesn't exist, create it
+      if (!CompositionBrush())
+      {
+        auto loadedSurface{ winrt::LoadedImageSurface::StartLoadFromUri(SourceUri()) };
+
+        loadedSurface.LoadCompleted([this](IInspectable const&, winrt::LoadedImageSourceLoadCompletedEventArgs const& args)
+        {
+          if (args.Status() == winrt::LoadedImageSourceLoadStatus::Success)
+          {
+            m_imageLoaded = true;
+            UpdateCompositionBrush();
+          }
+        });
+
+        auto surfaceBrush{ winrt::Window::Current().Compositor().CreateSurfaceBrush() };
+        surfaceBrush.Surface(loadedSurface);
+
+        return surfaceBrush;
+      }
+
+      auto surfaceBrush{ CompositionBrush().try_as<winrt::CompositionSurfaceBrush>() };
+
+      // If ResizeMode is set to Repeat, then we're using a CompositionEffectBrush.
+      // Get the CompositionSurfaceBrush from its source.
+      if (!surfaceBrush)
+      {
+        auto effectBrush{ CompositionBrush().as<winrt::CompositionEffectBrush>() };
+        surfaceBrush = effectBrush.GetSourceParameter(L"source").as<winrt::CompositionSurfaceBrush>();
+      }
+
+      return surfaceBrush;
+    }
+
+    bool ReactImageBrush::ShouldSwitchCompositionBrush()
+    {
+      bool effectToSurfaceBrushSwitch{
+        ResizeMode() != ResizeMode::Repeat &&
+        !CompositionBrush().try_as<winrt::CompositionSurfaceBrush>() };
+
+      bool surfaceToEffectBrushSwitch{ ResizeMode() == ResizeMode::Repeat };
```

Comment:
>bool surfaceToEffectBrushSwitch{ ResizeMode() == ResizeMode::Repeat }; [](start = 6, length = 70)

This seems to be missing a bit of logic (I think it should also check for a null m_effectBrush).

Reply:
The effect brush is created and/or retrieved just before the call to this method and the GetOrCreateEffectBrush call is also gated by the same logic (ResizeMode == Repeat). 

All other modes use the SurfaceBrush, so we always have to switch brushes if ResizeMode has been set to Repeat.

Reply:
Thanks for the explanation.  Reading through this a bit more carefully, I see where I got confused - this function isn't really checking to see if we need to switch brushes as it might already be set to the appropriate brush.  It looks like you could remove ShouldSwitchCompositionBrush() altogether, yes?  As it stands the code is somewhat confusing.

---
In reply to: [293640055](https://github.com/microsoft/react-native-windows/pull/2581#discussion_r293640055) [](ancestors = 293640055)

Reply:
It does check if we need to switch brushes. It's basically just preventing an unnecessary CompositionBrush set when we are switching between two resizeModes that use a CompositionSurfaceBrush (center, contain, cover, stretch). In those cases all we need to do is update CompositionSurfaceBrush.Stretch which is what we do right after we get the SurfaceBrush. 

I think the most human way of expressing this is in the negative (i.e. when do we NOT want to switch brushes), so maybe this is more readable?

bool ReactImageBrush::ShouldSwitchCompositionBrush()
{
    // The only scenario where we do not want to switch brushes is if:
    // 1. We're not switching to the EffectBrush (Repeat is the only resizeMode that uses the EffectBrush) and
    // 2. We're already using the SurfaceBrush
    return !(ResizeMode() != ResizeMode::Repeat && UsingSurfaceBrush())
}

---
In reply to: [294010205](https://github.com/microsoft/react-native-windows/pull/2581#discussion_r294010205) [](ancestors = 294010205,293640055)

Reply:
If I'm reading the code correctly, everything that leads up to the "should switch brushes" check is computing the brush that we should set.  If we were to always set this brush it would be the correct brush.  In other words, the "should switch brushes" check is just an optimization.

Assuming I've got all this correct, why can't you just query the current CompositionBrush() and compare to the one you're about to set, and only set if they are different?